base/generator.jl

```diff
@@ -52,7 +52,16 @@ size(g::Generator) = size(g.iter)
 axes(g::Generator) = axes(g.iter)
 ndims(g::Generator) = ndims(g.iter)
 
-getindex(g::Generator, I...) = map(g.f, g.iter[I...])
+function getindex(g::Generator, I...)
+    I′ = to_indices(g.iter, I)
```

Comment:
`to_indices` doesn't work on non-array indexables like `NamedTuple` and `Dict`. As I commented earlier, I don't think we should support vectorized indexing.

You can use LazyArrays.jl or MappedArrays.jl to get a full array API. If we were to add vectorized indexing for lazy mapping in `Base`, `Broadcasted` is probably a better interface to do it since it's already very array-like.

Reply:
> As I commented earlier, I don't think we should support vectorized indexing.

The tricky part of your statement is how would we limit `getindex` to just use scalar indexing?

Reply:
Just define `getindex(::Generator, ::Integer...)` and nothing else.  I also think this is the right thing to do.

Reply:
> how would we limit `getindex` to just use scalar indexing?

Ah, good point. How about introducing

```julia
function to_scalar_indices(A, I)
    J = to_indices(A, I)
    J isa Tuple{Vararg{Integer}} || error("expected scalar indices. got: ", I)
    return J
end

scalar_getindex(A::AbstractArray, I...) = getindex(A, to_scalar_indices(A, I)...)
scalar_getindex(A, I...) = getindex(A, I...)
```

and use it for `Generator`?

> Just define `getindex(::Generator, ::Integer...)`

Isn't it too restrictive and a half-way solution if it were to restrict the API? For example, it'd mean `x[1]` for `x = Iterators.map(string, (a=1, b=2))` works but `x[:a]` doesn't.

Also, I guess we need a different implementation for `AbstractDict` or error out? Applying `f` after `getindex` is not correct for `AbstractDict`.

Reply:
> Just define `getindex(::Generator, ::Integer...)`

Mostly that works but we get into some strange corner cases with `AbstractDict`:

```julia
julia> function Base.getindex(g::Base.Generator, I::Integer...)
           g.f(g.iter[I...])
       end

julia> g = (x * y for (x, y) in Dict(3 => 4))
Base.Generator{Dict{Int64,Int64},var"#1#2"}(var"#1#2"(),
Dict(3 => 4))

julia> collect(g)[1]
12

julia> g[1]
ERROR: KeyError: key 1 not found
Stacktrace:
 [1] getindex at ./dict.jl:467 [inlined]
 [2] getindex(::Base.Generator{Dict{Int64,Int64},var"#1#2"}, ::Int64) at ./REPL[1]:2
 [3] top-level scope at REPL[3]:1
```
This was the main reason I ended up restricting the logic to `AbstractArray`

Reply:
> I guess we need a different implementation for `AbstractDict` or error out

That was also my conclusion. A first pass at the `AbstractDict` version looks like:
```julia
function Base.getindex(g::Base.Generator{<:AbstractDict}, I...)
    subset = collect(pairs(g.iter))[I...]
    if subset isa Pair
        g.f(subset)
    else
        map(g.f, subset)
    end
end
```

Reply:
As in my new comment to this whole PR, why do you want to support AbstractDict in general?
IMO Generator should behave like a lazy map. And for map we have:
```julia
julia> map(x->println(x), Dict(:a=>2, :c=>"h"))
ERROR: map is not defined on dictionaries
Stacktrace:
 [1] error(::String) at .\error.jl:33
 [2] map(::Function, ::Dict{Symbol,Any}) at .\abstractarray.jl:2190
 [3] top-level scope at REPL[10]:1
```